native/src/main/java/io/ballerina/stdlib/data/FromString.java

```diff
@@ -0,0 +1,143 @@
+
+import io.ballerina.runtime.api.TypeTags;
+import io.ballerina.runtime.api.creators.ValueCreator;
+import io.ballerina.runtime.api.types.Type;
+import io.ballerina.runtime.api.types.UnionType;
+import io.ballerina.runtime.api.values.BDecimal;
+import io.ballerina.runtime.api.values.BString;
+import io.ballerina.runtime.api.values.BTypedesc;
+import io.ballerina.stdlib.data.utils.DataUtils;
+
+import java.util.Comparator;
+import java.util.List;
+
+/**
+ *  Native implementation of data:fromStringWithType(string).
+ * 
+ * @since 0.1.0
+ */
+public class FromString {
+
+    public static Object fromStringWithType(BString string, BTypedesc typed) {
+        Type expType = typed.getDescribingType();
+
+        try {
+            return fromStringWithType(string, expType);
+        } catch (NumberFormatException e) {
+            return DataUtils.getError("'string' value '" + string + "' cannot be converted to '" + expType + "'");
+        }
+    }
+
+    private static Object fromStringWithType(BString string, Type expType) {
+        String value = string.getValue();
+
+        switch (expType.getTag()) {
+            case TypeTags.INT_TAG:
+                return stringToInt(value);
+            case TypeTags.FLOAT_TAG:
+                return stringToFloat(value);
+            case TypeTags.DECIMAL_TAG:
+                return stringToDecimal(value);
+            case TypeTags.STRING_TAG:
+                return string;
+            case TypeTags.BOOLEAN_TAG:
+                return stringToBoolean(value);
+            case TypeTags.NULL_TAG:
+                return stringToNull(value);
+            case TypeTags.UNION_TAG:
+                return stringToUnion(string, (UnionType) expType);
+            default:
+                throw new IllegalStateException("Invalid type: " + expType);
+        }
+    }
+
+    private static Long stringToInt(String value) throws NumberFormatException {
+        return Long.parseLong(value);
+    }
+
+    private static Double stringToFloat(String value) throws NumberFormatException {
+        if (hasFloatOrDecimalLiteralSuffix(value)) {
+            throw new NumberFormatException();
+        }
+        return Double.parseDouble(value);
+    }
+
+    private static BDecimal stringToDecimal(String value) throws NumberFormatException {
+        return ValueCreator.createDecimalValue(value);
+    }
+    
+    private static Boolean stringToBoolean(String value) throws NumberFormatException {
+        if ("true".equalsIgnoreCase(value) || "1".equalsIgnoreCase(value)) {
+            return true;
+        } else if ("false".equalsIgnoreCase(value) || "0".equalsIgnoreCase(value)) {
+            return false;
+        }
+        throw new NumberFormatException();
+    }
+
+    private static Object stringToNull(String value) throws NumberFormatException {
+        if ("null".equalsIgnoreCase(value) || "()".equalsIgnoreCase(value)) {
+            return null;
+        }
+        throw new NumberFormatException();
```

Comment:
I think using `NumberFormatException` here is too arbitrary.

Reply:
Do you mean only for `stringToNull`?

Reply:
https://github.com/ballerina-platform/ballerina-lang/blob/0ec12efb0a4436dc85911c419c521a4c8b180b7e/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java#L981 This is an existing logic in the lang.

Reply:
In function which are not for handling number conversions such as `stringToBoolean` & `stringToUnion`

Reply:
Updated f3eb6dcbdbb4f80b0ed09b184702c48565e3212d